controllers/vmware/virtualmachinegroup_reconciler.go

```diff
@@ -0,0 +1,392 @@
+	"k8s.io/klog/v2"
+	clusterv1 "sigs.k8s.io/cluster-api/api/core/v1beta2"
+	"sigs.k8s.io/cluster-api/util/conditions"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	vmwarev1 "sigs.k8s.io/cluster-api-provider-vsphere/apis/vmware/v1beta1"
+	infrautilv1 "sigs.k8s.io/cluster-api-provider-vsphere/pkg/util"
+)
+
+const (
+	// ZoneAnnotationPrefix is the prefix used for placement decision annotations which will be set on VirtualMachineGroup.
+	ZoneAnnotationPrefix = "zone.vmware.infrastructure.cluster.x-k8s.io"
+)
+
+// VirtualMachineGroupReconciler reconciles VirtualMachineGroup.
+type VirtualMachineGroupReconciler struct {
+	Client   client.Client
+	Recorder record.EventRecorder
+}
+
+// +kubebuilder:rbac:groups=cluster.x-k8s.io,resources=clusters,verbs=get;list;watch
+// +kubebuilder:rbac:groups=cluster.x-k8s.io,resources=clusters/status,verbs=get
+// +kubebuilder:rbac:groups=vmoperator.vmware.com,resources=virtualmachinegroups,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=vmoperator.vmware.com,resources=virtualmachinegroups/status,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=vmware.infrastructure.cluster.x-k8s.io,resources=vspheremachines,verbs=get;list;watch
+// +kubebuilder:rbac:groups=cluster.x-k8s.io,resources=machinedeployments,verbs=get;list;watch
+// +kubebuilder:rbac:groups=cluster.x-k8s.io,resources=machines,verbs=get;list;watch
+
+// This controller is introduced by CAPV to coordinate the creation and maintenance of
+// the VirtualMachineGroup (VMG) object with respect to the worker VSphereMachines in the Cluster.
+//
+// - Batch Coordination: Gating the initial creation of the VMG until all expected worker
+// VSphereMachines are present. It will compare the desired count of VirtualMachines
+// derived from the sum of all desired MachineDeployment Replicas) against the count
+// of existing, matching VSphereMachine objects.
+//
+// - Placement Persistence: Persisting the MachineDeployment-to-Zone mapping (placement decision) as a
+// metadata annotation on the VMG object. The same decision must be respected also for placement
+// of machines created after initial placement.
+//
+// - Membership Maintenance: Dynamically updating the VMG's member list to reflect the current
+// state of VMs belonging to MachineDeployments (handling scale-up/down events).
+
+func (r *VirtualMachineGroupReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	// Fetch the Cluster instance.
+	cluster := &clusterv1.Cluster{}
+	if err := r.Client.Get(ctx, req.NamespacedName, cluster); err != nil {
+		if apierrors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	log = log.WithValues("VirtualMachineGroup", klog.KObj(cluster))
```

Comment:
```suggestion
	// Note: VirtualMachineGroup is going to have same name and namespace of the cluster.
	// Using cluster here, because VirtualMachineGroup is created only after initial placement completes.
	log = log.WithValues("VirtualMachineGroup", klog.KObj(cluster))
```